### ISSUE TITLE
IPFS data source

### ISSUE THREAD

**Issue by yanivtal** [graphprotocol/graph-node#262]
For bulk importing data from IPFS we need a way to support a new IPFS Import data source type.
The IPFS Import data source would point to an IPFS hash in the subgraph manifest and when the node begins indexing that subgraph for the first time it would bulk import that data only once.

**Comment by Jannis** [comment]
# Requirements

- Must be able to read large files (ENS; we can mandate line-based streaming).
- Must support JSON (either on the Graph Node side or in the mapping).
- Must be backwards-compatible (i.e., not break existing subgraph manifests or mappings).
- Must be resumable (for large files).

# Proposed solution

- Users define an IPFS data source in the subgraph manifest for bulk imports:

        dataSources:
        - kind: ipfs/file
          name: Districts
          source:
            file:
              /: /ipfs/QmP9DRyHqinwQrknCb7MkioDP7zmGaEAYrVL16wBg58xVt
          mapping:
            kind: ipfs/lines
            apiVersion: 0.0.1
            language: wasm/assemblyscript
            handler: handleLine
            file: ./src/ipfs-mapping.ts

- The data source includes a mapping that allows to read the file line by line and process the line (as `bytes` or `string`) using an AssemblyScript handler (satisfying the large file support).
- The handler can then call e.g. `json.fromBytes` or `json.fromString` on the line to parse it as JSON (satisfying the JSON requirement).
- After that, it is up to users to turn the JSON data from each line into entities.

# Required changes

1. Add a spec for an `ipfs/file` source and an `ipfs/lines` mapping.
    - The `ipfs/file` data source allows for a single IPFS file to be indexed.
    - The `ipfs/lines` mapping expects the IPFS file to be newline delimited UTF-8 and allows to stream the file one by one.
2. Graph CLI:
    - Update the GraphQL schema for validating subgraph manifests in Graph CLI.
    - Add validations for the `ipfs/file` data source and `ipfs/lines` mapping.
    - Include `ipfs/file` data sources and `ipfs/lines` mappings in the build and deploy process.
3. Graph Node:
    - Update the *subgraph of subgraphs* GraphQL schema.
    - Write meta entities for IPFS data sources and mappings to the subgraph of subgraphs.
    - Add support for unions to the GraphQL execution. This is needed to support different types of data sources / mappings. *(We can also use interfaces here but they won't have much in common beyond kind and name.)*
    - When starting a subgraph in the `SubgraphInstanceManager`
        - Only start a block stream if there any `ethereum/contract` data sources.
        - For every `ipfs/file` data source with an `ipfs/lines` mapping:
            1. Spawn an `IpfsFileLinesStream` to asynchronously read the lines of the IPFS file.
            2. Forward each line to the subgraph instance via something like `instance.process_ipfs_line(&logger, &link, line)`
        - Like block streams, IPFS line streams need to be stateful and resumable. They have to write the number of processed lines, processed bytes and total number of bytes to the database.
- Explorer API:
    - TODO
- Graph Explorer:
    - TODO
1. Misc:
    - Document IPFS data sources and mappings.
    - Nice to have: Add an IPFS data source to the example subgraph.

**Comment by lutter** [comment]
We now have `ipfs.map` that does what's outlined above (except for being resumable), and we solved the ENS bulk import differently.
